plaso/parsers/apthistory.py

```diff
@@ -0,0 +1,192 @@
+  """
+
+  DATA_TYPE = 'apthistory:log:line'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(AptHistoryLogEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.requestor = None
+    self.error = None
+    self.packages = None
+
+
+class AptHistoryLogParser(text_parser.PyparsingMultiLineTextParser):
+  """Parses events from APT History log files."""
+
+  NAME = 'apthistory'
+
+  DESCRIPTION = 'Parser for APT History log files.'
+
+  _ENCODING = 'utf-8'
+
+  # Increase the buffer size, as log messages can be very long.
+  BUFFER_SIZE = 131072
+
+  _HYPHEN = text_parser.PyparsingConstants.HYPHEN
+
+  _FOUR_DIGITS = text_parser.PyparsingConstants.FOUR_DIGITS
+  _TWO_DIGITS = text_parser.PyparsingConstants.TWO_DIGITS
+
+  _APTHISTORY_DATE_TIME = pyparsing.Group(
+      _FOUR_DIGITS + _HYPHEN +
+      _TWO_DIGITS + _HYPHEN +
+      _TWO_DIGITS +
+      _TWO_DIGITS + pyparsing.Suppress(':') +
+      _TWO_DIGITS + pyparsing.Suppress(':') +
+      _TWO_DIGITS
+  )
+
+  _APTHISTORY_LINE = (
+      pyparsing.ZeroOrMore(pyparsing.lineEnd()) +
+      pyparsing.Word('Start-Date: ').suppress() +
+      _APTHISTORY_DATE_TIME.setResultsName('start_date') +
+      pyparsing.SkipTo('End-Date: ').setResultsName('message') +
+      pyparsing.GoToColumn(10) +
+      _APTHISTORY_DATE_TIME.setResultsName('end_date') +
+      pyparsing.ZeroOrMore(pyparsing.lineEnd()))
+
+  LINE_STRUCTURES = [('logline', _APTHISTORY_LINE),]
+
+  def _ParseRecordLogline(self, parser_mediator, structure):
+    """Parses a logline record structure and produces events.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      structure (pyparsing.ParseResults): structure of tokens derived from
+          log entry.
+    """
+    time_zone = parser_mediator.timezone
+    time_elements_structure = self._GetValueFromStructure(
+        structure, 'start_date')
+    try:
+      date_time = dfdatetime_time_elements.TimeElements(
+          time_elements_tuple=time_elements_structure)
+      # APT History logs store date and time values in local time.
+      date_time.is_local_time = True
+    except ValueError:
+      parser_mediator.ProduceExtractionWarning(
+          'invalid date time value: {0!s}'.format(time_elements_structure))
+      return
+
+    event_data = AptHistoryLogEventData()
+
+    message = self._GetValueFromStructure(structure, 'message')
+    lines = message.split('\n')
+
+    for line in lines:
+      if 'Error:' in line:
+        event_data.error = line
+      if 'Requested-By:' in line:
+        event_data.requestor = line
+
+    for line in lines:
+      event_data.packages = line
+      if 'Downgrade:' in line:
+        event = time_events.DateTimeValuesEvent(
+            date_time,
+            definitions.TIME_DESCRIPTION_DOWNGRADE,
+            time_zone=time_zone)
+        parser_mediator.ProduceEventWithEventData(event, event_data)
```

Comment:
It looks like the idea is that only one of these strings appears in the line - if that's correct, please add a "continue" after the event and event_data are produced.

Reply:
This was deliberate. Not sure if it ever occurs, but my idea was that if more than one operation appears in a single log entry, I would create a separate event for each operation.